### PR TITLE
Build relationships of variables.

### PR DESCRIPTION
**Pull request by CPWstatic**

### CODE REVIEW

**Review by yixinglu — COMMENTED** [review]
two concerned points:
1. why not store the dependent relationships inside variable struct?
2. Maybe better to let plan node depend on QueryContext directly.

File: src/planner/PlanNode.h [hunk 1]
```diff
@@ -216,23 +234,57 @@ class SingleInputNode : public SingleDependencyNode {
...
+            }
+            qctx_->symTable()->deleteRead(oldVar, this);
+            qctx_->symTable()->readBy(inputVar, this);
+        } else {
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(inputVar, output->name);
+                qctx_->symTable()->addDependency(output->name, inputVar);
+            }
+            qctx_->symTable()->readBy(inputVar, this);
```
Comment (yixinglu) [inline src/planner/PlanNode.h:259]: Maybe we'd better to extract this logic to a `addInputVar` function so SingleInputNode also could use it.

File: src/planner/PlanNode.h [hunk 2]
```diff
@@ -216,23 +234,57 @@ class SingleInputNode : public SingleDependencyNode {
...
         if (dep != nullptr) {
-            inputVars_.emplace_back(dep->outputVar());
+            auto* inputVarPtr = dep->outputVarPtr();
+            inputVars_.emplace_back(inputVarPtr);
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(inputVarPtr->name, output->name);
+                qctx_->symTable()->addDependency(output->name, inputVarPtr->name);
+            }
+            qctx_->symTable()->readBy(inputVarPtr->name, this);
```
Comment (yixinglu) [inline src/planner/PlanNode.h:285]: ditto

File: src/planner/PlanNode.h [hunk 3]
```diff
@@ -248,11 +300,59 @@ class BiInputNode : public PlanNode {
...
+            qctx_->symTable()->deleteRead(oldVar, this);
+            qctx_->symTable()->readBy(leftVar, this);
+        } else {
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(leftVar, output->name);
+                qctx_->symTable()->addDependency(output->name, leftVar);
+            }
+            qctx_->symTable()->readBy(leftVar, this);
+        }
```
Comment (yixinglu) [inline src/planner/PlanNode.h:327]: ditto

File: src/planner/PlanNode.h [hunk 4]
```diff
@@ -248,11 +300,59 @@ class BiInputNode : public PlanNode {
...
+            }
+            qctx_->symTable()->deleteRead(oldVar, this);
+            qctx_->symTable()->readBy(rightVar, this);
+        } else {
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(rightVar, output->name);
+                qctx_->symTable()->addDependency(output->name, rightVar);
+            }
+            qctx_->symTable()->readBy(rightVar, this);
```
Comment (yixinglu) [inline src/planner/PlanNode.h:354]: ditto

File: src/planner/PlanNode.h [hunk 5]
```diff
@@ -264,33 +364,40 @@ class BiInputNode : public PlanNode {
...
+            qctx_->symTable()->addDependency(output->name, leftVarPtr->name);
         }
-
-        if (right != nullptr) {
-            dependencies_.emplace_back(right);
-            inputVars_.emplace_back(right->outputVar());
-        } else {
-            dependencies_.emplace_back();
-            inputVars_.emplace_back();
+        qctx_->symTable()->readBy(leftVarPtr->name, this);
```
Comment (yixinglu) [inline src/planner/PlanNode.h:390]: ditto

File: src/planner/PlanNode.h [hunk 6]
```diff
@@ -264,33 +364,40 @@ class BiInputNode : public PlanNode {
...
+
+        dependencies_.emplace_back(right);
+        auto* rightVarPtr = right->outputVarPtr();
+        inputVars_.emplace_back(rightVarPtr);
+        for (auto& output : outputVars_) {
+            DCHECK(output != nullptr);
+            qctx_->symTable()->addDerivative(rightVarPtr->name, output->name);
+            qctx_->symTable()->addDependency(output->name, rightVarPtr->name);
         }
+        qctx_->symTable()->readBy(rightVarPtr->name, this);
```
Comment (yixinglu) [inline src/planner/PlanNode.h:400]: ditto

File: src/planner/Query.h [hunk 7]
```diff
@@ -876,13 +875,24 @@ class DataCollect final : public SingleDependencyNode {
...
+        for (auto& var : vars) {
+            auto* inputVarPtr = qctx_->symTable()->getVar(var);
+            DCHECK(inputVarPtr != nullptr);
+            inputVars_.emplace_back(inputVarPtr);
+            for (auto& output : outputVars_) {
+                DCHECK(output != nullptr);
+                qctx_->symTable()->addDerivative(var, output->name);
+                qctx_->symTable()->addDependency(output->name, var);
+            }
+            qctx_->symTable()->readBy(var, this);
```
Comment (yixinglu) [inline src/planner/Query.h:894]: and this

File: src/planner/Query.h [hunk 8]
```diff
@@ -930,15 +940,39 @@ class DataJoin final : public SingleDependencyNode {
...
+
+        auto* leftVarPtr = qctx_->symTable()->getVar(leftVar_.first);
+        DCHECK(leftVarPtr != nullptr);
+        inputVars_.emplace_back(leftVarPtr);
+        for (auto& output : outputVars_) {
+            DCHECK(output != nullptr);
+            qctx_->symTable()->addDerivative(leftVar_.first, output->name);
+            qctx_->symTable()->addDependency(output->name, leftVar_.first);
+        }
+        qctx_->symTable()->readBy(leftVarPtr->name, this);
```
Comment (yixinglu) [inline src/planner/Query.h:964]: And here

File: src/planner/Query.h [hunk 9]
```diff
@@ -930,15 +940,39 @@ class DataJoin final : public SingleDependencyNode {
...
+
+        auto* rightVarPtr = qctx_->symTable()->getVar(rightVar_.first);
+        DCHECK(rightVarPtr != nullptr);
+        inputVars_.emplace_back(rightVarPtr);
+        for (auto& output : outputVars_) {
+            DCHECK(output != nullptr);
+            qctx_->symTable()->addDerivative(rightVar_.first, output->name);
+            qctx_->symTable()->addDependency(output->name, rightVar_.first);
+        }
+        qctx_->symTable()->readBy(rightVarPtr->name, this);
```
Comment (yixinglu) [inline src/planner/Query.h:974]: and here

**Review by yixinglu — APPROVED** [review]
Well Done. Thanks for your efforts.

**Review by dutor — APPROVED** [review]
LGTM.

**Pull request merged**